src/transformStateWithClones.js

```diff
@@ -6,8 +6,44 @@
  *
  * @return {Object[]}
  */
+function copyStateAndPush(Obj, arr) {
+  const copyObj = { ...Obj };
+
+  arr.push(copyObj);
+}
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const historyArray = [];
+  let copyState = { ...state };
+
+  for (const act in actions) {
+    if (actions[act].type === 'addProperties') {
+      const commandData = actions[act].extraData;
+
+      Object.assign(copyState, commandData);
+
+      copyStateAndPush(copyState, historyArray);
+      continue;
+    }
+
+    if (actions[act].type === 'removeProperties') {
+      const commandData = actions[act].keysToRemove;
+
+      for (const removeData of commandData) {
+        delete copyState[removeData];
+      }
+      copyStateAndPush(copyState, historyArray);
+      continue;
+    }
+
+    if (actions[act].type === 'clear') {
+      copyState = {};
+      copyStateAndPush(copyState, historyArray);
+      continue;
+    }
+  }
```

Comment:
Using a `switch/case` would be better than multiple `if`s on the same variable with well defined options